_posts/2020-02-23-pyflink-udf-support-flink.md

```diff
@@ -0,0 +1,280 @@
+* **Logging Service**: Mechanism for logging support for user defined functions. It allows transferring log entries produced by user defined functions to the Flink operator and integrates with Flink’s own logging system. 
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-02-23-pyflink-udfs/communication-flink-operator-python-execution-environment.png" width="600px" alt="Communication between the Flink operator and the Python execution environment"/>
+</center>
+<br>
+
+<div class="alert alert-info" markdown="1">
+<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Note</span>
+NOTE: Supporting metrics is currently planned for Flink 1.11
+</div>
+
+<br>
+
+Figure 2 below describes the high-level flow between initializing and executing UDFs from the Java operator to the Python process. 
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-02-23-pyflink-udfs/flow-initialization-execution-python-udfs-flink.png" width="800px" alt="High-level flow between initialization and execution of Python UDFs in Flink 1.10"/>
+</center>
+<br>
+
+The high-level flow can be summarized in two parts:
+1. Initialization of the Python execution environment.
+  * The Python UDF Runner starts the gRPC services, such as the data service, logging service, etc.
+  * The Python UDF Runner launches the Python execution environment in a separate process.
+  * The Python worker registers to `PythonUserDefinedFunctionRunner`.
+  * The Python UDF Runner sends the user defined functions to be executed in the Python worker.
+  * The Python worker transforms the user defined functions to Beam operations. (Note that we leverage the power of [Beam’s Portability Framework](https://beam.apache.org/roadmap/portability/) to execute the Python UDF.)
+  * The Python worker establishes the gRPC connections, such as the data connection, logging connection, etc.
+
+2. Processing of the input elements.
+  * The Python UDF Runner sends the input elements via the gRPC data service to the Python worker for execution.
+  * The Python use defined function can access the state via gRPC state service during execution.
+  * The Python user defined function can also aggregate the logging and metrics to the Python UDF Runner via the gRPC logging service and the metrics service during execution.
+  * The execution results are finally sent to the Python UDF Runner via the gRPC data service.
+
+
+# How to use PyFlink with UDFs in Flink 1.10
+
+This section provides some Python user defined function (UDF) examples, including how to install PyFlink, how to define/register/invoke UDFs in PyFlink and how to execute the job.
+
+
+### Install PyFlink
+Using Python in Apache Flink requires installing PyFlink. PyFlink is available through PyPi and can be easily installed using pip: 
+
+```python
```

Comment:
The installation commands are missing here (and maybe something else?)

Reply:
Add a heading for how to use Python UDFs

Reply:
Maybe first show a simple but complete Pyflink program without a UDF?

Reply:
> The installation commands are missing here (and maybe something else?)

+1, I think the command `python -m pip install apache-flink`  should be added  .  same as `Install PyFlink` part in [1].
[1] https://docs.google.com/document/d/1I83o8e4CSu1baKPK9SKq1S1GN-p98AVvDOIgUHQVOZo/edit# 

> Add a heading for how to use Python UDFs

This post is talk about Python UDF support, So I am fine without a simple example. :)